### PR TITLE
Deprecate synced flush

### PR DESCRIPTION
**Pull request by dnhatn**
A normal flush has the same effect as a synced flush on Elasticsearch 7.6 or later. It's deprecated in 7.6 and will be removed in 8.0.

Relates #50776

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 1]
```diff
@@ -423,7 +423,10 @@ public Cancellable flushAsync(FlushRequest flushRequest, RequestOptions options,
      * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
      * @return the response
      * @throws IOException in case there is a problem sending the request or parsing back the response
+     * @deprecated synced flush is deprecated and will be removed in future Elasticsearch versions.
+     * Please use normal {@link #flush(FlushRequest, RequestOptions)} instead.
```
Comment (jrodewig) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]:
```suggestion
     * Use {@link #flush(FlushRequest, RequestOptions)} instead.
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 2]
```diff
@@ -437,7 +440,10 @@ public SyncedFlushResponse flushSynced(SyncedFlushRequest syncedFlushRequest, Re
      * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
      * @param listener the listener to be notified upon request completion
      * @return cancellable that may be used to cancel the request
+     * @deprecated synced flush is deprecated and will be removed in future Elasticsearch versions.
+     * Please use normal {@link #flushAsync(FlushRequest, RequestOptions, ActionListener)} instead.
```
Comment (jrodewig) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]:
```suggestion
     * Use {@link #flushAsync(FlushRequest, RequestOptions, ActionListener)} instead.
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 3]
```diff
@@ -437,7 +440,10 @@ public SyncedFlushResponse flushSynced(SyncedFlushRequest syncedFlushRequest, Re
      * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
      * @param listener the listener to be notified upon request completion
      * @return cancellable that may be used to cancel the request
+     * @deprecated synced flush is deprecated and will be removed in future Elasticsearch versions.
```
Comment (jrodewig) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]:
```suggestion
     * @deprecated Synced flush is deprecated and will be removed in a future version.
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java [hunk 4]
```diff
@@ -1018,7 +1018,10 @@ public void testSyncedFlushIndex() throws Exception {
             // end::flush-synced-request-indicesOptions
 
             // tag::flush-synced-execute
-            SyncedFlushResponse flushSyncedResponse = client.indices().flushSynced(request, RequestOptions.DEFAULT);
+            SyncedFlushResponse flushSyncedResponse = client.indices().flushSynced(request, expectWarnings(
+                "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
```
Comment (jrodewig) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]:
```suggestion
                "Synced flush is deprecated and will be removed in a future version. " +
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java [hunk 5]
```diff
@@ -1018,7 +1018,10 @@ public void testSyncedFlushIndex() throws Exception {
...
+                "Please use normal flush at _/flush or /{index}/_flush instead."
```
Comment (jrodewig) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]:
```suggestion
                "Use _/flush or /{index}/_flush instead."
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java [hunk 6]
```diff
@@ -1062,7 +1065,10 @@ public void onFailure(Exception e) {
             listener = new LatchedActionListener<>(listener, latch);
 
             // tag::flush-synced-execute-async
-            client.indices().flushSyncedAsync(request, RequestOptions.DEFAULT, listener); // <1>
+            client.indices().flushSyncedAsync(request, expectWarnings(
+                "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
+                "Please use normal flush at _/flush or /{index}/_flush instead."
```
Comment (jrodewig) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]:
```suggestion
                "Use _/flush or /{index}/_flush instead."
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java [hunk 7]
```diff
@@ -1062,7 +1065,10 @@ public void onFailure(Exception e) {
             listener = new LatchedActionListener<>(listener, latch);
 
             // tag::flush-synced-execute-async
-            client.indices().flushSyncedAsync(request, RequestOptions.DEFAULT, listener); // <1>
+            client.indices().flushSyncedAsync(request, expectWarnings(
+                "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
```
Comment (jrodewig) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/IndicesClientDocumentationIT.java]:
```suggestion
                "Synced flush is deprecated and will be removed in future a version. " +
```

File: docs/reference/indices/synced-flush.asciidoc [hunk 8]
```diff
@@ -12,6 +12,10 @@ POST /twitter/_flush/synced
 --------------------------------------------------
 // TEST[skip: Synced flush can conflict with scheduled flushes in doc tests]
 
+deprecated:[7.6, synced-flush is deprecated and will be removed in future
```
Comment (jrodewig) [inline docs/reference/indices/synced-flush.asciidoc]:
```suggestion
deprecated:[7.6, Synced flush is deprecated and will be removed in a future
```

File: docs/reference/indices/synced-flush.asciidoc [hunk 9]
```diff
@@ -12,6 +12,10 @@ POST /twitter/_flush/synced
...
+Elasticsearch versions. Please use normal <<indices-flush,flush>> instead.
```
Comment (jrodewig) [inline docs/reference/indices/synced-flush.asciidoc]:
```suggestion
version. Use <<indices-flush,flush>> instead.
```

File: docs/reference/indices/synced-flush.asciidoc [hunk 10]
```diff
@@ -12,6 +12,10 @@ POST /twitter/_flush/synced
...
+A normal <<indices-flush,flush>> has the same effect as a synced flush on
```
Comment (jrodewig) [inline docs/reference/indices/synced-flush.asciidoc]:
```suggestion
Flush has the same effect as synced flush in
```

File: docs/reference/indices/synced-flush.asciidoc [hunk 11]
```diff
@@ -12,6 +12,10 @@ POST /twitter/_flush/synced
...
+Elasticsearch 7.6 or later]
```
Comment (jrodewig) [inline docs/reference/indices/synced-flush.asciidoc]:
```suggestion
{es} 7.6 and later versions.]
```

File: docs/reference/upgrade/synced-flush.asciidoc [hunk 12]
```diff
@@ -3,8 +3,13 @@
...
 POST _flush/synced
 --------------------------------------------------
+// TEST[skip: will fail as synced flush is deprecated]
 
 When you perform a synced flush, check the response to make sure there are
 no failures. Synced flush operations that fail due to pending indexing
 operations are listed in the response body, although the request itself
 still returns a 200 OK status. If there are failures, reissue the request.
+
+Note that synced flush is deprecated and will be removed in future versions.
```
Comment (jrodewig) [inline docs/reference/upgrade/synced-flush.asciidoc]:
```suggestion
NOTE: Synced flush is deprecated and will be removed in a future version.
```

File: docs/reference/upgrade/synced-flush.asciidoc [hunk 13]
```diff
@@ -3,8 +3,13 @@
...
+A normal flush has the same effect as a synced flush on Elasticsearch 7.6
```
Comment (jrodewig) [inline docs/reference/upgrade/synced-flush.asciidoc]:
```suggestion
Flush has the same effect as synced flush in {es} 7.6
```

File: docs/reference/upgrade/synced-flush.asciidoc [hunk 14]
```diff
@@ -3,8 +3,13 @@
...
+or later.
```
Comment (jrodewig) [inline docs/reference/upgrade/synced-flush.asciidoc]:
```suggestion
and later versions.
```

File: rest-api-spec/src/main/resources/rest-api-spec/api/indices.flush_synced.json [hunk 15]
```diff
@@ -2,7 +2,7 @@
   "indices.flush_synced":{
     "documentation":{
       "url":"https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-synced-flush-api.html",
-      "description":"Performs a synced flush operation on one or more indices."
+      "description":"Performs a synced flush operation on one or more indices. Synced flush is deprecated and will be removed in future Elasticsearch versions. Use normal flush instead"
```
Comment (jrodewig) [inline rest-api-spec/src/main/resources/rest-api-spec/api/indices.flush_synced.json]:
```suggestion
      "description":"Performs a synced flush operation on one or more indices. Synced flush is deprecated and will be removed in a future version. Use flush instead."
```

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.flush/10_basic.yml [hunk 16]
```diff
@@ -1,26 +1,33 @@
...
-    - do:
-       indices.flush_synced:
-          index: testing
+  - do:
+      cluster.health:
+        wait_for_status: green
+  - do:
+      warnings:
+        - Synced flush is deprecated and will be removed in future Elasticsearch versions.
+          Please use normal flush at _/flush or /{index}/_flush instead.
```
Comment (jrodewig) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.flush/10_basic.yml]:
```suggestion
          Please use  _/flush or /{index}/_flush instead.
```

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.flush/10_basic.yml [hunk 17]
```diff
@@ -1,26 +1,33 @@
...
-          wait_for_status: green
-    - do:
-       indices.flush_synced:
-          index: testing
+  - do:
+      cluster.health:
+        wait_for_status: green
+  - do:
+      warnings:
+        - Synced flush is deprecated and will be removed in future Elasticsearch versions.
```
Comment (jrodewig) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.flush/10_basic.yml]:
```suggestion
        - Synced flush is deprecated and will be removed in a future version.
```

File: server/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportSyncedFlushAction.java [hunk 18]
```diff
@@ -32,7 +34,14 @@
...
 public class TransportSyncedFlushAction extends HandledTransportAction<SyncedFlushRequest, SyncedFlushResponse> {
 
-    SyncedFlushService syncedFlushService;
+    private final SyncedFlushService syncedFlushService;
+
+    private static final DeprecationLogger DEPRECATION_LOGGER =
+        new DeprecationLogger(LogManager.getLogger(TransportSyncedFlushAction.class));
+
+    public static final String SYNCED_FLUSH_DEPRECATION_MESSAGE =
+        "Synced flush is deprecated and will be removed in future Elasticsearch versions. " +
```
Comment (jrodewig) [inline server/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportSyncedFlushAction.java]:
```suggestion
        "Synced flush is deprecated and will be removed in a future version. " +
```

File: server/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportSyncedFlushAction.java [hunk 19]
```diff
@@ -32,7 +34,14 @@
...
+            "Please use normal flush at _/flush or /{index}/_flush instead.";
```
Comment (jrodewig) [inline server/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportSyncedFlushAction.java]:
```suggestion
            "Please use _/flush or /{index}/_flush instead.";
```

**Review by jrodewig — COMMENTED** [review]
Hi @dnhatn!

Overall, I think the docs for this are in pretty good shape. I left several non-blocking nits throughout.

The only blocking change I requested is re: the `deprecated::[]` note in `synced-flush.asciidoc`. The current deprecation block isn't rendering properly. I'd also move it higher up on the page.

**Comment by dnhatn** [comment]
Hi @jrodewig Thank you for reviewing. I've addressed your comments.

Discussed with @ywelsch on another channel, we prefer this PR make to 7.x only so we can write the docs more concise.

**Review by jrodewig — APPROVED** [review]
Docs LGTM. Thanks @dnhatn!

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks James and Yannick!

**Pull request merged**